contracts/protocol/Insurance/Insurance.sol

```diff
@@ -0,0 +1,103 @@
+pragma solidity ^0.4.11;
+
+import "../Utils/Ownable.sol";
+import "../Agreement/EscrowedAgreement.sol";
+
+contract Insurance is Ownable {
+
+    enum Status {
+        Unknown,
+        Active,
+        ReimbursementRequested,
+        ReimbursementApproved,
+        ReimbursementRejected,
+        Complete
+    }
+
+    struct InsuranceDetails {
+        Status status;
+        address beneficiary;
+    }
+
+    ERC20Protocol private token;
+
+    mapping (address => InsuranceDetails) public insuredAgreements;
+
+    event NewAgreementInsured(address beneficiary, address agreement);
+    event ReimbursementRequested(address requestor, address agreement);
+    event ReimbursementApproved(address agreement);
+    event ReimbursementRejected(address agreement);
+    event ReimbursementPaid(address payee, address agreement, uint amount);
+
+    function Insurance (ERC20Protocol _token) {
+        token = _token;
+    }
+
+    function () public {
+        // Nothing. Do not accept ETH transfers.
+    }
+
+    function insure(EscrowedAgreement agreement, uint premium) public {
+        require (agreement != address(0x0));
+        require (insuredAgreements[agreement].status == Status.Unknown);
+        require (agreement.seller() == msg.sender || agreement.buyer() == msg.sender);
+        require (premium > 0);
+
+        /* 
+            The specific insurance contracts inheriting from this contract must override 
+            this function to do additional validations of agreement, premium and other details
+            before underwriting.
+        */
+
+        insuredAgreements[agreement] = InsuranceDetails(Status.Active, msg.sender);
+        
+        assert(token.transferFrom(msg.sender, this, premium));
```

Comment:
This could be perhaps `require`, as it depends on external conditions (i.e. user called approve on the token), not an internal error. Also could be perhaps moved above the previous line (to save money to a user if he calls accidentally without calling `approve`).

Reply:
Done